Schema

ManicJamie opened the pull request:
Create a JSON Schema for validation.

## Justification

JSON Schema are a standard for validation JSON, YAML and TOML files with extensive pre-existing tooling available. The current `schemaChecks.py` script already uses `jsonschema` in its function, but crucially includes custom logic & isn't too readable itself.

A unified schema will make format change proposals easier & allow greater interoperability.

## Format Changes

I made a few changes to simplify the final schema, including;
- `release-year` removed, to be replaced with `release-date` (currently omitted)
- `info-updated` -> `updated-date` to be in line with `added-date`
- most nullable entries made non-nullable but instead as optional keys, as the JSON schema should indicate their existence
- Dates at the "import epoch" given a clearer default of `1900-01-01`

ManicJamie commented:
Formatting changes will require minor fixes to generation & a new action.

Die4Ever commented:
This is good yea. Although minor regression, we no longer support conditional requirements based on updated date. Right now this means new entries no longer require the opensource field here. It also means adding new fields will be more difficult, we'd need to add the field, transform all the data, and update the templates, all at the same time. Our data is stable enough now that it probably isn't a big deal anymore.

release-date instead of release-year might be a little annoying, I've seen disagreements on exact dates between Wikipedia, IGDB, MobyGames, IMDB, etc. The first country/platform of release might also confuse some people when a different release date is more popular. But right now we're only showing the release year anyways, so the full release date would be dormant.

updated-date sounds a lot like when the randomizer most recently released a patch, instead of when we update our info for it. I wouldn't want to confuse submitters.

You didn't include my changes for Pokemon.yml from the other open PR :( I hope you still have the code for making the transformations.

I did a spot check and you have Deus_Ex.yml with an updated-date of 1900-01-01, but it should be 2024-06-19

Die4Ever commented on src/series/Deus_Ex.yml:
```diff
@@ -1,17 +1,20 @@
     identifier: Die4Ever's
     url: https://mods4ever.com/project/DXRando
-    comment: "Supports the regular vanilla game and also the mods: GMDX, HX, Lay D Denton, Revision, and Vanilla? Madder."
+    comment: 'Supports the regular vanilla game and also the mods: GMDX, HX, Lay D
```
Minor complaint but when you did the transformation, the max width on your output was maybe a bit small?

Die4Ever commented on src/series/Deus_Ex.yml:
```diff
@@ -1,17 +1,20 @@
     opensource: true
-    added-date: 2024-05-25
-    info-updated: 2024-06-19
+    updated-date: 1900-01-01
```
this lost the updated-date it had before

Die4Ever commented on validate-schema.py:
```diff
@@ -0,0 +1,39 @@
+            try:
+                validate(series[filename], schema, cls=Validator)
+            except Exception as e:
+                logging.error(e, exc_info=True)
```
`exc_info=True` makes this pretty verbose for more casual submitters? also it doesn't always print the filename when outputting errors, like if I change genre to RPGaaaa

```suggestion
                print('ERROR in', filename.name)
                logging.error(e)
```

Die4Ever commented on validate-schema.py:
```diff
@@ -0,0 +1,39 @@
+    series = {}
+    for filename in Path("src/series/").glob("*"):
+        with open(filename, encoding="utf-8") as f:
+            series[filename] = yaml.safe_load(f)
```
we used to have filename validation, like

```
if re.search(r'[^a-zA-Z0-9_\-\.\(\)]', filename.stem):
    raise RuntimeError("bad filename " + filename.name)
```

Die4Ever replied:
if a file is invalid yml then it blows up the whole script instead of moving on to the next file, we should scoot this into the try block

Die4Ever commented:
this doesn't seem to decline invalid dates, such as `added-date: 1900-01-01324`

Die4Ever commented:
another minor regression, this only tells 1 error per file, but we might not be able to fix that with this method

Die4Ever commented on src/schemata/series.schema.json:
```diff
@@ -0,0 +1,142 @@
+            "description": "Date this randomizer was added",
+            "type": "string",
+            "format": "date",
+            "default": "1900-01-01"
```
I don't think we need this default since we filled them all in, and new entries should definitely have it. I removed this line and it still succeeded for me.


```suggestion
```

ManicJamie replied:
`default` in JSON-schema is not a 'null values get coerced to this', but rather a 'missing values are semantically equivalent to this'. Afaik no validators utilise `default` at all. In fact, the spec [only recommends it even be a valid value](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-validation-00#rfc.section.9.2)

Die4Ever commented:
"most nullable entries made non-nullable but instead as optional keys, as the JSON schema should indicate their existence"

I think early on we decided against this because more casual submitters might not even realize all the fields they could include

but I'm fine with it either way

Die4Ever commented:
Another minor regression: we no longer check for missing games or unused games. Again not sure if we can fix that with this setup, but we could still do some post-processing in the python code for that or for date checks.

ManicJamie commented:
Trying to collect all in one comment;

## Resolved
- Output width
- Dates missing
- Rebased on origin/main

## Discussion

### Starting with the things I believe are non-issues;

> Although minor regression, we no longer support conditional requirements based on updated date. Right now this means new entries no longer require the opensource field here.

Schema are for automatic format validation, not to fully describe submission details. Submission-date-specific schema is an odd requirement & can just as easily be achieved by listing the required attributes for new submissions - we need to manually verify these submissions regardless.

> It also means adding new fields will be more difficult, we'd need to add the field, transform all the data, and update the templates, all at the same time. Our data is stable enough now that it probably isn't a big deal anymore.

IMO keeping operations like this atomic per-PR is best-practice & not a major hurdle - I can track the migration script I made for this if desired. Making all data fully coherent/compliant should be a short-term goal.

> this doesn't seem to decline invalid dates, such as added-date: `1900-01-01324`

This appears to be an issue with the `format` field in general when it comes to JSON schema - Format checking is opt-in and `jsonschema`'s solution doesn't appear to work too well. Can add it as a lowpriority issue, but ultimately any dates added should be checked manually anyway.

> another minor regression, this only tells 1 error per file, but we might not be able to fix that with this method

Seems unfixable but ultimately if there are several format errors that indicates a more major issue that isn't solely relying on the validation action - the schema is public specifically so that contributors can use validators within their IDEs before even committing. Of course many contributors will likely use the Github web interface, but at that point a complete template in the README is really the best option.

### Open to further discussion

> "most nullable entries made non-nullable but instead as optional keys, as the JSON schema should indicate their existence"
> I think early on we decided against this because more casual submitters might not even realize all the fields they could include

This is a fair distinction; I chose this mainly for the readability of the schema, but honestly I'm not too sure on it myself. Additionally I now don't think readability is even a real concern.

> Another minor regression: we no longer check for missing games or unused games. Again not sure if we can fix that with this setup, but we could still do some post-processing in the python code for that or for date checks.

Maybe useful - might be worth adding a separate `validate-data` step to the action.

> updated-date sounds a lot like when the randomizer most recently released a patch, instead of when we update our info for it. I wouldn't want to confuse submitters.

I think this is also solveable through proper documentation, but I admit making the schema less self-documenting is a meaningful regression for the minor benefit of `"created-date" and "info-updated" are in the opposite order :(`

> release-date instead of release-year might be a little annoying, I've seen disagreements on exact dates between Wikipedia, IGDB, MobyGames, IMDB, etc. The first country/platform of release might also confuse some people when a different release date is more popular. But right now we're only showing the release year anyways, so the full release date would be dormant.

I added full release date for easier sorting; the exact dates should not be critical here since it should be mostly used to sort games within a series. For that use case I think the potential footpistols are minimal, but I could be proven wrong.

Die4Ever commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 src/_site
 src/Gemfile.lock
+respell-schema.py
+.vscode/
```
what's in your .vscode/? some of those things we might want in the repo

keeping the respell-schema.py might be useful as an example? maybe put the date in front of the filename?

ManicJamie replied:
respell-schema.py just iterates over Series -> Game & Randomizer, calling a function on each. Expectation is new requirements will require new code anyway :P 

.vscode/ contains workspace configs specific to a YAML schema validation plugin - I don't think they belong here, but extension/config could be linked in the contribution section of the README if it's really wanted

Die4Ever replied:
I mean yea but an example code might help people? idk, I mean I've done transformations before and they always end up really simple and specific but it could help someone?

.vscode/ can contain an extensions.json to suggest that extension, and the settings for it, I think it's a good idea to include and I believe that's a big reason why VSCode supports these files in the workspace/repo

like for DXRando we have extensions.json:

```
{
	// See https://go.microsoft.com/fwlink/?LinkId=827846 to learn about workspace recommendations.
	// Extension identifier format: ${publisher}.${name}. Example: vscode.csharp

	// List of extensions which should be recommended for users of this workspace.
	"recommendations": [
		"EditorConfig.EditorConfig", "EricBlade.vscode-unrealscript"
	],
	// List of extensions recommended by VS Code that should not be recommended for users of this workspace.
	"unwantedRecommendations": [
        "EliotVU.uc"
	]
}
```

where `EliotVU.uc` is an old extension for UnrealScript that doesn't work, and we recommend the other extension for UnrealScript instead, it helps people getting set up

Die4Ever commented on src/_includes/series.html:
```diff
@@ -17,6 +17,7 @@
     {%- assign game_data = include.series.games[firstGame] -%}
+    {%- assign release_year = game_data.release_date | split: "-" | first -%}
     {%- if game_data.release_year != null -%}
         {%- assign release_year_suffix = ' <span class="text-muted fs-5">(RELEASE)</span>' | replace: "RELEASE", game_data.release_year -%}
```
```suggestion
        {%- assign release_year_suffix = ' <span class="text-muted fs-5">(RELEASE)</span>' | replace: "RELEASE", release_year -%}
```

Die4Ever replied:
there's actually a few occurrences for release_year

![image](https://github.com/user-attachments/assets/da3aea81-a31d-4094-96ca-d983a9144316)

Die4Ever commented on validate-schema.py:
```diff
@@ -0,0 +1,43 @@
+    series = {}
+    for filename in Path("src/series/").glob("*"):
+        try:
+            if re.search(r'[^a-zA-Z0-9_\-\.\(\)]', filename.stem):
```
should also check the file extension, since the stem ignores that

```suggestion
            if filename.suffix != '.yml' or re.search(r'[^a-zA-Z0-9_\-\.\(\)]', filename.stem):
```

Die4Ever commented on validate-schema.py:
```diff
@@ -0,0 +1,43 @@
+    for filename in Path("src/series/").glob("*"):
+        try:
+            if re.search(r'[^a-zA-Z0-9_\-\.\(\)]', filename.stem):
+                raise RuntimeError("bad filename " + filename.name)
```
probably ValueError is better here

```suggestion
                raise ValueError("bad filename " + filename.name)
```

ok I'm done nitpicking lol, after these fixes I can approve this
